sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisServiceImpl.java

```diff
@@ -0,0 +1,160 @@
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.redis;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.values.KV;
+
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.Pipeline;
+import redis.clients.jedis.Response;
+import redis.clients.util.JedisClusterCRC16;
+
+/**
+ * An implementation of the {@link RedisService} that actually communicates with a Redis server.
+ */
+public class RedisServiceImpl implements RedisService {
+
+  private final RedisConnectionConfiguration connection;
+
+  public RedisServiceImpl(RedisConnectionConfiguration connection) {
+    this.connection = connection;
+  }
+
+  private class RedisReaderImpl extends BoundedSource.BoundedReader<KV<String, String>> {
+
+    private final RedisIO.RedisSource source;
+
+    private Jedis jedis;
+    private Iterator<String> keysIterator;
+    private KV<String, String> current;
+
+    public RedisReaderImpl(RedisIO.RedisSource source) {
+      this.source = source;
+    }
+
+    @Override
+    public boolean start() {
+      if (source.node != null && source.node.host != null) {
+        jedis = new Jedis(source.node.host, source.node.port, connection.timeout());
+      } else {
+        jedis = connection.connect();
+      }
+      Pipeline pipeline = jedis.pipelined();
+      Response<Set<String>> keysResponse;
+      if (source.keyPattern == null) {
+        keysResponse = pipeline.keys("*");
+      } else {
+        keysResponse = pipeline.keys(source.keyPattern);
+      }
+      pipeline.syncAndReturnAll();
+
+      Set<String> keys = keysResponse.get();
+
+      keysIterator = keys.iterator();
+      return advance();
+    }
+
+    @Override
+    public boolean advance() {
+      if (keysIterator.hasNext()) {
+        String key = keysIterator.next();
+        String value = jedis.get(key);
+        KV<String, String> kv = KV.of(key, value);
+        current = kv;
+        return true;
+      }
+      return false;
+    }
+
+    @Override
+    public void close() {
+      jedis.quit();
+    }
+
+    @Override
+    public KV<String, String> getCurrent() throws NoSuchElementException {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public BoundedSource<KV<String, String>> getCurrentSource() {
+      return source;
+    }
+
+  }
+
+  @Override
+  public RedisReaderImpl createReader(RedisIO.RedisSource source) {
+    return new RedisReaderImpl(source);
+  }
+
+  /**
+   * The estimate size bytes is based on sampling, computing average size of 10 random
+   * key/value pairs. This sampling average size is used with the Redis dbSize to get an
+   * estimation of the actual database size.
+   */
+  @Override
+  public long getEstimatedSizeBytes() {
```

Comment:
This seems likely to be very inaccurate, because it ignores the key pattern and it ignores the fact that the database may be storing many different kinds of keys, with wildly different value size distributions.

Rather than asking to make this more accurate, I'm asking to reconsider whether perhaps it's better to implement RedisIO.Read as a DoFn, and implement size estimation when somebody asks for it. I think, given how hard it is to make it accurate and how little it matters in practice, it's likely to be not worth the complexity.